[SPARK-22074][Core] Task killed by other attempt task should not be resubmitted

xuanyuanking opened the pull request:
## What changes were proposed in this pull request?

As the detail scenario described in [SPARK-22074](https://issues.apache.org/jira/browse/SPARK-22074), unnecessary resubmitted may cause stage hanging in currently release versions. This patch add a new var in TaskInfo to mark this task killed by other attempt or not.

## How was this patch tested?

Add a new UT `[SPARK-22074] Task killed by other attempt task should not be resubmitted` in TaskSetManagerSuite, this UT recreate the scenario in JIRA description, it failed without the changes in this PR and passed conversely.

xuanyuanking committed:
[SPARK-22074][Core] Task killed by other attempt task should not be resubmitted

SparkQA commented:
**[Test build #81969 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81969/testReport)** for PR 19287 at commit [`a28daa2`](https://github.com/apache/spark/commit/a28daa2c3283ad31659f840e6d401ab48a42ad88).

SparkQA commented:
**[Test build #81969 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81969/testReport)** for PR 19287 at commit [`a28daa2`](https://github.com/apache/spark/commit/a28daa2c3283ad31659f840e6d401ab48a42ad88).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81969/
Test FAILed.

xuanyuanking commented:
`signal 9`
retest this please

SparkQA commented:
**[Test build #81987 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81987/testReport)** for PR 19287 at commit [`a28daa2`](https://github.com/apache/spark/commit/a28daa2c3283ad31659f840e6d401ab48a42ad88).

SparkQA commented:
**[Test build #81987 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81987/testReport)** for PR 19287 at commit [`a28daa2`](https://github.com/apache/spark/commit/a28daa2c3283ad31659f840e6d401ab48a42ad88).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81987/
Test PASSed.

markhamstra commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -66,6 +66,12 @@ class TaskInfo(
+  /**
+   * Set this tag when this task killed by other attempt. This kind of task should not resubmit
+   * while executor lost.
+   */
```
nit: calling it a "tag" doesn't aid understanding, but rather leaves me wondering why it was called a "tag" instead of a "var" or something more descriptive.

more consequential: Please explain what defines "this kind" of task.

xuanyuanking replied:
Thanks for reviewing, rewrite this comment.

xuanyuanking committed:
Comment rewrite

SparkQA commented:
**[Test build #82018 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82018/testReport)** for PR 19287 at commit [`81ac4dc`](https://github.com/apache/spark/commit/81ac4dc75f91c888a9fefa805915f9420d71b761).

SparkQA commented:
**[Test build #82018 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82018/testReport)** for PR 19287 at commit [`81ac4dc`](https://github.com/apache/spark/commit/81ac4dc75f91c888a9fefa805915f9420d71b761).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82018/
Test PASSed.

xuanyuanking commented:
ping @cloud-fan @gatorsmile

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -66,6 +66,13 @@ class TaskInfo(
+   * set the `spark.speculation` to true. The task killed by others should not resubmit
+   * while executor lost.
+   */
+  var killedAttempt = false
```
killedByAttemp

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -93,6 +104,8 @@ class TaskInfo(
 
   def running: Boolean = !finished
 
+  def needResubmit: Boolean = !killedAttempt
```
this looks unnecessary, we can just write `!killedByAttempt`

markhamstra replied:
Since `killedAttempt`/`killedByAttempt` is never used except in this negated sense, it probably makes more sense to get rid of it and this `needResubmit` method entirely, and just have a `needResubmit` var with the opposite sense of `killedAttempt` -- but I'm also not all that happy with the `needResubmit` name.

cloud-fan commented:
@jerryshao

xuanyuanking committed:
Simplify code and nit fix

SparkQA commented:
**[Test build #82176 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82176/testReport)** for PR 19287 at commit [`039591d`](https://github.com/apache/spark/commit/039591d4fb2cbe3292b4f0ce33ba605bed895453).

SparkQA commented:
**[Test build #82176 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82176/testReport)** for PR 19287 at commit [`039591d`](https://github.com/apache/spark/commit/039591d4fb2cbe3292b4f0ce33ba605bed895453).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82176/
Test FAILed.

jerryshao commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -66,6 +66,13 @@ class TaskInfo(
+   * set the `spark.speculation` to true. The task killed by others should not resubmit
+   * while executor lost.
+   */
+  var killedByAttempt = false
```
I would suggest to rename to `killedByOtherAttempt`.

jerryshao commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -74,6 +81,10 @@ class TaskInfo(
     gettingResultTime = time
   }
 
+  private[spark] def markKilledAttempt {
```
It is better to change the method signature to `def markKilledByOtherAttempt(): Unit = {`, since this method has side affect, it is better to add parenthesis.

jerryshao reviewed (commented):
Hi @xuanyuanking , I left some questions on JIRA, currently I'm not fully understanding the issue here, would you please comment on the JIRA, thanks!

xuanyuanking committed:
Rename the variable

SparkQA commented:
**[Test build #82192 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82192/testReport)** for PR 19287 at commit [`7191eff`](https://github.com/apache/spark/commit/7191eff18dcad8f680964b2cb8df8c68c27de801).

SparkQA commented:
**[Test build #82192 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82192/testReport)** for PR 19287 at commit [`7191eff`](https://github.com/apache/spark/commit/7191eff18dcad8f680964b2cb8df8c68c27de801).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82192/
Test PASSed.

jerryshao commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -74,6 +81,10 @@ class TaskInfo(
     gettingResultTime = time
   }
 
+  private[spark] def markKilledAttempt: Unit = {
```
I think I suggested you to add parenthesis in this method signature `def markKilledByOtherAttempt()`, can you please change it?

> It is better to change the method signature to def markKilledByOtherAttempt(): Unit = {, since this method has side affect, it is better to add parenthesis.

xuanyuanking replied:
Sorry for the missing, I change it right now.

xuanyuanking committed:
Change method signature and rename

SparkQA commented:
**[Test build #82262 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82262/testReport)** for PR 19287 at commit [`71b0d58`](https://github.com/apache/spark/commit/71b0d5821ca4d2738fb608073acbb8f0ba1d8d29).

SparkQA commented:
**[Test build #82262 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82262/testReport)** for PR 19287 at commit [`71b0d58`](https://github.com/apache/spark/commit/71b0d5821ca4d2738fb608073acbb8f0ba1d8d29).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82262/
Test PASSed.

jerryshao commented:
Generally it looks fine to me.

CC @markhamstra @squito , would you please help to review it? Thanks!

xuanyuanking commented:
@jerryshao Thanks for you review.

squito commented on core/src/test/scala/org/apache/spark/scheduler/FakeTask.scala:
```diff
@@ -58,4 +57,18 @@ object FakeTask {
   }
+
+  def createShuffleMapTaskSet(numTasks: Int, stageId: Int, stageAttemptId: Int,
+    prefLocs: Seq[TaskLocation]*): TaskSet = {
```
nit: multi-line method definitions should have each param on its own line, indented 4 spaces

```scala
def createShuffleMapTaskSet(
    numTasks: Int,
    stageId: Int,
    stageAttemptId: Int,
    prefLocs: Seq[TaskLocation]*): TaskSet = {
```

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -744,6 +744,100 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+      "exec1" -> "host1",
+      "exec1" -> "host1",
+      "exec3" -> "host3",
+      "exec2" -> "host2")) {
```
nits:
rename `k` -> `exec`
`v` -> `host`

double indent the body of `List`

in general, prefer `Seq` to `List` (not really a big deal, but since you are touching this anyway...)

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -744,6 +744,100 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+    assert(task4.taskId === 4)
+    assert(task4.executorId === "exec2")
+    assert(task4.attemptNumber === 1)
+    sched.backend = mock(classOf[SchedulerBackend])
```
its really weird to switch `sched.backend` in the middle of the test.  I worry that in the future this will not work as expected and the test will not work as expected when something else changes in the scheduler.

instead, can you just override `killTask` in your subclass of `FakeSchedulerBackend` to track the calls you care about?

xuanyuanking replied:
Yep, in this case there's only one `killTask` check, fixed this in next patch.

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -744,6 +744,100 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+      val task = taskOption.get
+      assert(task.executorId === k)
+    }
+    assert(sched.startedTasks.toSet === Set(0, 1, 2, 3))
```
can you also add an assert somewhere around here that task 2.0 is running on exec3?  I see that locality preference should give you that, but its critical for the final part of the test so worth an extra assert.

xuanyuanking replied:
OK, I add the assert while get the `taskOption`

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -66,6 +66,13 @@ class TaskInfo(
+   * set the `spark.speculation` to true. The task killed by others should not resubmit
+   * while executor lost.
+   */
+  var killedByOtherAttempt = false
```
`TaskInfo` is a little funny since its DeveloperApi.  Its exposed to the user via SparkListener events.  If want to leave it publicly visible, then I think there should only be a getter, not a public `var`, and also we'd need to update the way its converted to / from json in `o.a.s.util.JsonProtocol.taskInfoToJson` etc.  But I think its perhaps better to just mark this as `private[scheduler]`, as I dont' think its actually important for the end user.

squito replied:
actually, I had another thought -- rather than putting this info in `TaskInfo`, I think its more natural to add something totally private to `TaskSetManager`:

`private val killedByOtherAttempt: Array[Boolean]`

xuanyuanking replied:
Thanks to point out the `TaskInfo` is DeveloperApi, I address this as your second comment, add this totally private to TaskSetManager

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskInfo.scala:
```diff
@@ -74,6 +81,10 @@ class TaskInfo(
     gettingResultTime = time
   }
 
+  private[spark] def markKilledByOtherAttempt(): Unit = {
```
can tighten visibility to `private[scheduler]`

squito reviewed (changes requested):
Thanks a lot for this @xuanyuanking -- great job documenting how it occurs, creating a test case and fixing.  All my comments are pretty minor things.

xuanyuanking committed:
Move the killedByOtherAttempt into TaskSetManager and fix other nits

SparkQA commented:
**[Test build #82306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82306/testReport)** for PR 19287 at commit [`94fd257`](https://github.com/apache/spark/commit/94fd257c3a8da6ef4473eab72e826af57b10ed47).

xuanyuanking commented:
@squito Hi Rashid, thanks for you review and advise. In the last commit I moved `killedByOtherAttempt` into `TaskSetManager ` as you say and added more asserts in UT.

SparkQA commented:
**[Test build #82306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82306/testReport)** for PR 19287 at commit [`94fd257`](https://github.com/apache/spark/commit/94fd257c3a8da6ef4473eab72e826af57b10ed47).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82306/
Test FAILed.

xuanyuanking commented:
retest this please

SparkQA commented:
**[Test build #82313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82313/testReport)** for PR 19287 at commit [`94fd257`](https://github.com/apache/spark/commit/94fd257c3a8da6ef4473eab72e826af57b10ed47).

SparkQA commented:
**[Test build #82313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82313/testReport)** for PR 19287 at commit [`94fd257`](https://github.com/apache/spark/commit/94fd257c3a8da6ef4473eab72e826af57b10ed47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82313/
Test PASSed.

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -744,6 +744,112 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+    manager.handleSuccessfulTask(4, createTaskResult(2, accumUpdatesByTask(2)))
+    // With this successful task end, the sched.backend will kill other running attempt,
+    // verify the request of killTask(2, "exec3", true, "another attempt succeeded") in
+    // FakeDAGScheduler subclass
```
this is *almost* as good as your previous code involving the mock, but there is one scenario its different -- you no longer make sure that `killTask` actually gets called at all.  I think you should have it set a flag which you check here.

That additional assert also makes the comment unnecessary (or if you want to keep it, can be more succinct, eg. "Make sure schedBackend.killTask(2, "exec3", true, "another attempt succeeded") gets called"

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -744,6 +744,112 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+       taskId: Long,
+       executorId: String,
+       interruptThread: Boolean,
+       reason: String): Unit = {
```
style nit:  each arg in the method defintion should be indented 4 spaces, looks like their indented only 1 space here.

squito reviewed (approved):
a couple very minor comments, otherwise lgtm

xuanyuanking committed:
Nit fix

SparkQA commented:
**[Test build #82343 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82343/testReport)** for PR 19287 at commit [`5d09b3b`](https://github.com/apache/spark/commit/5d09b3ba410ca69b3c67d219ff2a4ad9998db7d8).

SparkQA commented:
**[Test build #82343 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82343/testReport)** for PR 19287 at commit [`5d09b3b`](https://github.com/apache/spark/commit/5d09b3ba410ca69b3c67d219ff2a4ad9998db7d8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82343/
Test FAILed.

xuanyuanking committed:
Fix IDE style change

SparkQA commented:
**[Test build #82349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82349/testReport)** for PR 19287 at commit [`1c8c849`](https://github.com/apache/spark/commit/1c8c84937e85302f2ac48bcbdbdb5507c9b445e4).

SparkQA commented:
**[Test build #82349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82349/testReport)** for PR 19287 at commit [`1c8c849`](https://github.com/apache/spark/commit/1c8c84937e85302f2ac48bcbdbdb5507c9b445e4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82349/
Test PASSed.

squito commented:
lgtm, thanks @xuanyuanking 

@jerryshao can you merge this?  I will have very intermittent access for a few weeks, I'd prefer not to merge in case there is any issue that needs an urgent followup.

jerryshao commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #82546 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82546/testReport)** for PR 19287 at commit [`1c8c849`](https://github.com/apache/spark/commit/1c8c84937e85302f2ac48bcbdbdb5507c9b445e4).

SparkQA commented:
**[Test build #82546 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82546/testReport)** for PR 19287 at commit [`1c8c849`](https://github.com/apache/spark/commit/1c8c84937e85302f2ac48bcbdbdb5507c9b445e4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82546/
Test PASSed.

jerryshao commented:
LGTM, merging to master. Thanks!

xuanyuanking commented:
Thanks all reviewers!
